tests/util/Components.js

```diff
@@ -19,12 +21,32 @@ const ruleTester = new eslint.RuleTester({
 
 describe('Components', () => {
   describe('static detect', () => {
-    function testComponentsDetect(test, done) {
-      const rule = Components.detect((context, components, util) => ({
-        'Program:exit'() {
-          done(context, components, util);
-        },
-      }));
+    function testComponentsDetect(test, instructionsOrDone, orDone) {
+      const done = orDone || instructionsOrDone;
+      const instructions = orDone ? instructionsOrDone : instructionsOrDone;
+
+      const rule = Components.detect((_context, components, util) => {
+        const instructionResults = [];
+
+        const augmentedInstructions = fromEntries(
+          entries(instructions || {}).map((nodeTypeAndHandler) => {
+            const nodeType = nodeTypeAndHandler[0];
+            const handler = nodeTypeAndHandler[1];
+            return [nodeType, (node) => {
+              instructionResults.push({ type: nodeType, result: handler(node, context, components, util) });
```

Comment:
The purpose of the `Components` test scaffolding is to allow exfiltration of the `Components.detect` locals to a scope in the test where `assert` was available, but without having to write a complete rule with a `Program:exit` clause.

The original implementation exfiltrated the `Components.detect` locals, but didn't make it easy to actually test some of those locals; in particular I couldn't test `utils` because the utility functions expect `ASTNodes`, and by the time `Program:exit` fires, I've lost the `CallExpression` nodes I need.

These changes are intended to aid in writing new test cases for `Components` functionality, and in particular, for testing `utils` functionality, and those utils may expect many different ASTNode types, not just `CallExpression`.

Custom instructions passed into the `testComponentsDetect` get augmented and then incorporated into a custom rule, and when that rule runs over code the augmented instructions do two things:
1. Provide the `Components.detect` locals to the instruction callback. Ordinarily the instruction callback only receives a `node` as an argument and relies on having been defined in the `Components.detect` closure to access these local.
2. Accumulate a "log" as the visitor traverses the parsed AST; the log tracks which node type was traversed, and the result of that visit.

Taken together, these allow creation test cases intended to read something like:

> The visitor visited a `CallExpression` node
> The visitor had access to the `util` local
> The `util` function `isReactHookCall` returned the following result for the visited `CallExpression` node